### ISSUE TITLE
Optional `Api::get`

### ISSUE THREAD

**Issue by teozkr** [kube-rs/kube-rs#806]
### Would you like to work on this feature?

yes

### What problem are you trying to solve?

Sometimes I want to get an object if it exists, but have some fallback behaviour otherwise. For example, something along the lines of:

```rust
if let Some(obj) = cms.get_opt("foo").await? {
    Ok(obj)
} else {
    cms.create(generate_foo()).await
}
```

Today, this requires me to pattern-match on the `Error::Api`, and match the string `reason` against a constant. This is non-obvious (nothing indicates to new users that this is a viable thing to do), bug-prone (I get no warning if I make a typo in the reason), and interacts poorly with the borrow checker.

### Describe the solution you'd like

Introduce a new `Api::get_opt` (or `try_get` or `get_if_exists`?) that does this pattern matching internally, and returns a `Result<Option<K>>`, returning `Some(None)` for 404 errors.

### Describe alternatives you've considered

1. Modify `Api::get` rather than introducing a new method
   - Smaller API surface
   - Would disrupt existing users
   - Most use-cases would probably just want to fail loudly anyway
2. Introduce a new `Error::is_not_found(&self)` (or `ErrorResponse::is_not_found`?) method
   - More extensible for introducing new cases later
   - Still not very obvious
   - Would be usable in a lot of places where it's not really useful (especially if put on `Error`)
   - Not very ergonomic
3. Introduce a new `Error::NotFound` case
   - Would silently break anyone currently trying to match on `ErrorResponse.reason`

### Target crate for feature

kube-client

**Comment by clux** [comment]
This makes sense to me. My initial gut feel here was that we ought to change `Api::get` to return a `Result<Option<K>>` with us doing the 404 error matching directly... but adding a `get_opt` is more sensible from a breaking perspective, and it's the only sensible way for users who expect certainty to be able to use `.get` unobtrusively, and also have users who expect uncertainty to call something else (`.get_opt`).

..it's also maybe possible to mimic the [map Entry api](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html) here, or do a `Api::get_or_else`, but both of those feel a bit crazy.

**Comment by teozkr** [comment]
`Entry` or `get_or_insert` might actually make sense for the ger-or-create use-case...

**Comment by clux** [comment]
we could sketch out an `Api::entry -> kube::api::Entry` thing if you prefer to make something proper to cater for these situations

- `.or_default` makes sense for structs that have `Default` impls
- `.or_apply_with` might make sense instead of `.or_insert_with` / `.or_insert_with_key`
- `.or_apply` probably makes less sense 
- `.and_modify` i can see being useful

**Comment by imp** [comment]
I find myself quite a lot in situations where I need to check for an object to be present and do something else if it isn't (erroring out on failure for any other reason). Perhaps a combinator like `.found() -> kube::Result<Option<K>>` can be implemented to make this pattern clearer.
Something like
```
if let Some(obj) = Api::<K>::get().found()? {
    // Yey
} else {
   // No dice
}
```

That might involve creating an extension trait for `kube::Result`

Or perhaps even better idea: `Api::find() -> kube::Result<Option<K>>` to mirror the behaviour of `.find` for slices and iterators

**Comment by teozkr** [comment]
@imp Does #809 satisfy your requirements?

**Comment by imp** [comment]
Yeah, that's exactly that. The name `try_get` wouldn't be my first choice, because in other places in Rust `try_`  is a `Result<T, U>` counterpart of otherwise infallible operation. However, in `kube` almost everything is `kube::Result<T>`, so it may hide the intention. Other than that - cool!
